src/lib/structures/objective.ml

```diff
@@ -0,0 +1,118 @@
+(*     This file is distributed under the terms of OCamlPro               *)
+(*     Non-Commercial Purpose License, version 1.                         *)
+(*                                                                        *)
+(*     As an exception, Alt-Ergo Club members at the Gold level can       *)
+(*     use this file under the terms of the Apache Software License       *)
+(*     version 2.0.                                                       *)
+(*                                                                        *)
+(*     ---------------------------------------------------------------    *)
+(*                                                                        *)
+(*     The Alt-Ergo theorem prover                                        *)
+(*                                                                        *)
+(*     Sylvain Conchon, Evelyne Contejean, Francois Bobot                 *)
+(*     Mohamed Iguernelala, Stephane Lescuyer, Alain Mebsout              *)
+(*                                                                        *)
+(*     CNRS - INRIA - Universite Paris Sud                                *)
+(*                                                                        *)
+(*     Until 2013, some parts of this code were released under            *)
+(*     the Apache Software License version 2.0.                           *)
+(*                                                                        *)
+(*     ---------------------------------------------------------------    *)
+(*                                                                        *)
+(*     More details can be found in the directory licenses/               *)
+(*                                                                        *)
+(**************************************************************************)
+
+module Function = struct
+  type t = {
+    e : Expr.t;
+    to_max : bool;
+  }
+
+  let equal { e = e1; _ } { e = e2; _ } = Expr.equal e1 e2
+
+  let mk ~to_max e = { e; to_max }
+
+  let pp ppf { e; _ } = Expr.print ppf e
+end
+
+module Value = struct
+  type limit_kind =
+    | Above
+    | Below
+
+  type t =
+    | Minfinity
+    | Pinfinity
+    | Value of Expr.t
+    | Limit of limit_kind * Expr.t
+    | Unknown
+
+  let pp ppf v =
+    match v with
+    | Minfinity -> Fmt.pf ppf "-oo"
+    | Pinfinity -> Fmt.pf ppf "+oo"
+    | Value w -> Expr.print ppf w
+    | Limit (Above, w) -> Fmt.pf ppf "(+ %a epsilon)" Expr.print w
+    | Limit (Below, w) -> Fmt.pf ppf "(- %a epsilon)" Expr.print w
+    | Unknown -> Fmt.pf ppf "(interval (- oo) (+ oo))"
+end
+
+module Model = struct
+  type s = { f: Function.t; order: int }
+
+  module M =  Map.Make (struct
+      type t = s
+
+      let compare { order = o1; _ } { order = o2; _ } = o1 - o2
+    end)
+
+  type t = Value.t M.t
+
+  let empty = M.empty
+  let is_empty = M.is_empty
+  let fold g = M.fold (fun { f; _ } acc -> g f acc)
+
+  let add o v mdl =
```

Comment:
Nit: I was slightly confused by calling a function `o` (would have expected an order), why not `g`?

Reply:
I called it `o` because it's an `objective function`. Maybe `obj`?